cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/DeletedProgramHandlerStage.java

```diff
@@ -91,35 +89,13 @@ public void process(ApplicationDeployable appSpec) throws Exception {
       ProgramType type = ProgramTypes.fromSpecification(spec);
       final ProgramId programId = appSpec.getApplicationId().program(type, spec.getName());
       programTerminator.stop(programId);
+      programScheduler.deleteSchedules(programId);
+      programScheduler.modifySchedulesTriggeredByDeletedProgram(programId);
 
-      // TODO: Unify with AppFabricHttpHandler.removeApplication
       // drop all queues and stream states of a deleted flow
       if (ProgramType.FLOW.equals(type)) {
-        FlowSpecification flowSpecification = (FlowSpecification) spec;
-
-        // Collects stream name to all group ids consuming that stream
-        final Multimap<String, Long> streamGroups = HashMultimap.create();
-        for (FlowletConnection connection : flowSpecification.getConnections()) {
-          if (connection.getSourceType() == FlowletConnection.Type.STREAM) {
-            long groupId = FlowUtils.generateConsumerGroupId(programId, connection.getTargetName());
-            streamGroups.put(connection.getSourceName(), groupId);
-          }
-        }
-        // Remove all process states and group states for each stream
-        final String namespace = String.format("%s.%s", programId.getApplication(), programId.getProgram());
-
-        final NamespaceId namespaceId = appSpec.getApplicationId().getParent();
-        impersonator.doAs(appSpec.getApplicationId(), new Callable<Void>() {
-
-          @Override
-          public Void call() throws Exception {
-            for (Map.Entry<String, Collection<Long>> entry : streamGroups.asMap().entrySet()) {
-              streamConsumerFactory.dropAll(namespaceId.stream(entry.getKey()), namespace, entry.getValue());
-            }
-            queueAdmin.dropAllForFlow(programId.getParent().flow(programId.getEntityName()));
-            return null;
-          }
-        });
+        FlowUtils.clearDeletedFlow(impersonator, queueAdmin, streamConsumerFactory, programId,
```

Comment:
Unified with `ApplicationLifecycleService#deleteApp` by calling the same util method